File: backend/app/schemas/documents.py
from datetime import datetime
from typing import Any, Optional

from pydantic import BaseModel, ConfigDict, Field


class DocumentOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    filename: str
    stored_path: str
    content_type: str
    sha256: str
    status: str
    space_id: int
    space_name: str = ""
    char_count: int
    error_message: Optional[str] = None
    created_at: datetime
    updated_at: datetime


class SourceChunkOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    document_id: int
    space_id: int
    chunk_index: int
    title: str
    text: str
    start_char: int
    end_char: int
    page_start: Optional[int] = None
    page_end: Optional[int] = None
    section: str = ""
    clause_ids_json: str = "[]"
    parent_index: Optional[int] = None
    created_at: datetime


class RechunkOut(BaseModel):
    document_id: int
    chunk_count: int


class DocumentDeleteOut(BaseModel):
    document_id: int
    chunks_deleted: int = 0
    pages_archived: list[str] = Field(default_factory=list)
    pages_detached: list[str] = Field(default_factory=list)
    reviews_closed: int = 0
    source_file_removed: bool = False
    warnings: list[str] = Field(default_factory=list)


class DocumentPreviewOut(BaseModel):
    document_id: int
    filename: str
    text: str
    char_count: int
    returned_chars: int
    truncated: bool = False
    quality_ok: bool = True
    diagnostics: dict[str, Any] = Field(default_factory=dict)

File: frontend/src/api/documents.ts
import { api } from './client'

export interface DocumentItem {
  id: number
  filename: string
  stored_path: string
  content_type: string
  sha256: string
  status: string
  char_count: number
  error_message: string | null
  created_at: string
  updated_at: string
}

export interface DocumentDiagnostics {
  is_empty?: boolean
  replacement_char_count?: number
  replacement_char_rate?: number
  garbled_char_count?: number
  garbled_char_rate?: number
  suspicious_scanned_pdf?: boolean
  page_count?: number
  pages_with_text?: number
  warnings?: string[]
  errors?: string[]
  [key: string]: unknown
}

export interface DocumentPreview {
  document_id: number
  filename: string
  text: string
  char_count: number
  returned_chars: number
  truncated: boolean
  quality_ok: boolean
  diagnostics: DocumentDiagnostics
}

export interface IngestJob {
  id: number
  document_id: number
  status: string
  stage: string
  progress: number
  plan_json: string
  cancel_requested: boolean
  step_log_json: string
  error_message: string | null
  created_at: string
  updated_at: string
}

export interface IngestStep {
  step?: string
  message?: string
  at?: string
  progress?: number
  [key: string]: unknown
}

export interface SourceChunk {
  id: number
  document_id: number
  chunk_index: number
  title: string
  text: string
  start_char: number
  end_char: number
  page_start: number | null
  page_end: number | null
  section: string
  clause_ids_json: string
  parent_index: number | null
  created_at: string
}

export function listDocuments() {
  return api<DocumentItem[]>('/api/documents')
}

export function uploadDocument(file: File) {
  const form = new FormData()
  form.append('file', file)
  return api<DocumentItem>('/api/documents', {
    method: 'POST',
    body: form,
  })
}

export function getDocumentPreview(id: number, maxChars = 50000) {
  return api<DocumentPreview>(`/api/documents/${id}/preview?max_chars=${maxChars}`)
}

export function listDocumentChunks(id: number) {
  return api<SourceChunk[]>(`/api/documents/${id}/chunks`)
}

export function ingestDocument(id: number, force = false) {
  const query = force ? '?force=true' : ''
  return api<IngestJob>(`/api/documents/${id}/ingest${query}`, { method: 'POST' })
}

export function listIngestJobs(status?: string) {
  const query = status ? `?status=${encodeURIComponent(status)}` : ''
  return api<IngestJob[]>(`/api/ingest-jobs${query}`)
}

export function getIngestJob(id: number) {
  return api<IngestJob>(`/api/ingest-jobs/${id}`)
}

export function cancelIngestJob(id: number) {
  return api<IngestJob>(`/api/ingest-jobs/${id}/cancel`, { method: 'POST' })
}

export function retryFailedWindows(id: number) {
  return api<IngestJob>(`/api/ingest-jobs/${id}/retry-failed-windows`, { method: 'POST' })
}
